core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -542,7 +583,79 @@ public GroupByIdBlock getResult()
             checkState(lastPosition == block.getPositionCount(), "process has not yet finished");
             checkState(!finished, "result has produced");
             finished = true;
-            return new GroupByIdBlock(nextGroupId, blockBuilder.build());
+            return new GroupByIdBlock(nextGroupId, new LongArrayBlock(blockData.length, Optional.empty(), blockData));
+        }
+    }
+
+    private void batchedPutIfAbsent(int batchSize, int lastPosition, Block block, long[] groupIds, int groupIdOffset)
+    {
+        if (block.mayHaveNull()) {
+            boolean haveNull = false;
+
+            int[] hashes = new int[batchSize];
+            for (int i = 0; i < batchSize; i++) {
+                if (!block.isNull(lastPosition + i)) {
+                    hashes[i] = getHashPosition(BIGINT.getLong(block, lastPosition + i), mask);
+                }
+            }
+
+            for (int i = 0; i < batchSize; i++) {
+                if (!block.isNull(lastPosition + i)) {
+                    groupIds[groupIdOffset + i] = this.groupIds[hashes[i]];
+                }
+                else {
+                    groupIds[groupIdOffset + i] = -2;
+                    haveNull = true;
+                }
+            }
+
+            for (int i = 0; i < batchSize; i++) {
+                if (groupIds[groupIdOffset + i] >= 0) {
+                    long value = BIGINT.getLong(block, lastPosition + i);
+                    long storedValue = values[hashes[i]];
```

Comment:
for a large number of groups, this is a cache miss. The same is true for the `this.groupIds[hashes[i]]` above.
Since this loop is not super tight, manually unrolling could help here (I have seen improvement with doing 4 items at once)